src/node-file-trace.ts

```diff
@@ -1,17 +1,42 @@
 import { NodeFileTraceOptions, NodeFileTraceResult, NodeFileTraceReasons, Stats, NodeFileTraceReasonType } from './types';
-import { basename, dirname, extname, relative, resolve, sep } from 'path';
+import { basename, dirname, extname, relative, resolve, sep, join } from 'path';
+import * as url from 'url';
 import fs from 'graceful-fs';
 import analyze, { AnalyzeResult } from './analyze';
 import resolveDependency, { NotFoundError } from './resolve-dependency';
 import { isMatch } from 'micromatch';
 import { sharedLibEmit } from './utils/sharedlib-emit';
-import { join } from 'path';
 import { Sema } from 'async-sema';
 
 const fsReadFile = fs.promises.readFile;
 const fsReadlink = fs.promises.readlink;
 const fsStat = fs.promises.stat;
 
+type ParseSpecifierResult = {
+  path: string;
+  queryString: string | null
+}
+
+// Splits an ESM specifier into path and querystring (including the leading `?`). (If the specifier is CJS,
+// it is passed through untouched.)
+export function parseSpecifier(specifier: string, cjsResolve: boolean = true): ParseSpecifierResult {
+  let path = specifier;
+  let queryString = null;
+
+  if (!cjsResolve) {
+    const specifierUrl = url.parse(specifier);
+    queryString = specifierUrl.search;
+
+    if (specifierUrl.search) {
+      path = specifier.replace(specifierUrl.search, '');
```

Comment:
This could replace the wrong part of the URL if it had repeating content in the string. 

How about using the parsed result instead?

```suggestion
      path = specifierUrl.pathname;
```

Reply:
Because we want to exclusively strip the query string off the `specifier`. If we use `pathname` we would also strip away suff like hash, protocol, host etc. (I know it is a bit edge-casey but still, let's do as little as possible to the specifier here)

I will rename the function to something like `splitQueryStringFromSpecifier` and change the return object signature to `{ queryString, remainingSpecifier }` to make its purpose more clear.

Reply:
As for the repeating content, do you have examples of where this might go wrong?

Afaik, as per URL standard, the first `?` will denote the start of the query/`search` part of the URL. `search` should always start with a `?` and since `.replace()` will only replace the first occurrence of a match (unless something like the regex `g` flag is used) this should just work IMO.

Reply:
> we would also strip away stuff like hash, protocol, host etc.

Did you add tests for those cases? How does Node.js handle those cases?

>  since .replace() will only replace the first occurrence of a match

Ah yes, you're right.

> new URL("../../src/mymodule", "file://").pathname === "/src/mymodule")

Oh good catch!